### PR TITLE
Try to improve literal types

### PR DESCRIPTION
**Pull request by vaithak**
This is related to the review comments of PR #975.

### CODE REVIEW

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 1]
```diff
@@ -1008,8 +1008,11 @@ StmtDiff BaseForwardModeVisitor::VisitDeclRefExpr(const DeclRefExpr* DRE) {
   // If DRE is of type pointer, then the derivative is a null pointer.
   if (clonedDRE->getType()->isPointerType())
     return StmtDiff(clonedDRE, nullptr);
+  QualType literalTy = utils::GetValueType(clonedDRE->getType());
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1011]: We should extend visitorbase::getZeroInit as discussed here https://github.com/vgvassilev/clad/pull/989#discussion_r1684487454

cc: @gojakuch

**Comment by vgvassilev** [comment]
This pull request, if the review comment is addressed, should fix #1001.

**Comment by kchristin22** [comment]
The cuda bug was a result of not handling the enum types correctly. `clad::utils::GetValueType()` doesn't return the underlying type of the enum but the enum itself (as a type). When checking `isRealType()` for that node, the cudaError_t enum passes the test so it's type is not set as an Int. This lead to a compilation error down the line. 

I fixed that part in `clad::utils::GetValueType()` in a branch based on Vaibhav's `cuda-bug` branch. Should I open a new PR and reference this one and continue addressing the comments of this and its previous PR there?

**Comment by vgvassilev** [comment]
> The cuda bug was a result of not handling the enum types correctly. `clad::utils::GetValueType()` doesn't return the underlying type of the enum but the enum itself (as a type). When checking `isRealType()` for that node, the cudaError_t enum passes the test so it's type is not set as an Int. This lead to a compilation error down the line.
> 
> I fixed that part in `clad::utils::GetValueType()` in a branch based on Vaibhav's `cuda-bug` branch. Should I open a new PR and reference this one and continue addressing the comments of this and its previous PR there?

Can you push to this PR?

**Comment by kchristin22** [comment]
> > The cuda bug was a result of not handling the enum types correctly. `clad::utils::GetValueType()` doesn't return the underlying type of the enum but the enum itself (as a type). When checking `isRealType()` for that node, the cudaError_t enum passes the test so it's type is not set as an Int. This lead to a compilation error down the line.
> > I fixed that part in `clad::utils::GetValueType()` in a branch based on Vaibhav's `cuda-bug` branch. Should I open a new PR and reference this one and continue addressing the comments of this and its previous PR there?
> 
> Can you push to this PR?

Sure, but I have no permissions in Vaibhav's repo, @vaithak could you change it temporarily(till this PR is merged)?

**Comment by vgvassilev** [comment]
If you accept my invite you should be able to push to this PR.

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 2]
```diff
@@ -1070,8 +1070,11 @@ StmtDiff BaseForwardModeVisitor::VisitDeclRefExpr(const DeclRefExpr* DRE) {
...
+  if (!literalTy->isRealType())
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1074]: Why do we need this special check here?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 3]
```diff
@@ -1374,8 +1377,12 @@ StmtDiff BaseForwardModeVisitor::VisitUnaryOperator(const UnaryOperator* UnOp) {
   } else if (opKind == UnaryOperatorKind::UO_Deref) {
     if (Expr* dx = diff.getExpr_dx())
       return StmtDiff(op, BuildOp(opKind, dx));
-    return StmtDiff(op, ConstantFolder::synthesizeLiteral(
-                            m_Context.IntTy, m_Context, /*val=*/0));
+    QualType literalTy =
+        utils::GetValueType(UnOp->getSubExpr()->getType()->getPointeeType());
+    if (!literalTy->isRealType())
+      literalTy = m_Context.IntTy;
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1383]: I think we should not need this. This gets handled in `ConstantFolder::synthesizeLiteral` anyways.

File: lib/Differentiator/VisitorBase.cpp [hunk 4]
```diff
@@ -379,11 +379,10 @@ namespace clad {
     // FIXME: Consolidate other uses of synthesizeLiteral for creation 0 or 1.
     if (T->isVoidType())
       return nullptr;
-    if (T->isScalarType()) {
+    if (T->isScalarType() && !T->isPointerType()) {
```
Comment (vgvassilev) [inline lib/Differentiator/VisitorBase.cpp:382]: Why do we need this here? Shouldn't `ConstantFolder::synthesizeLiteral` check if T is a pointer type and synthesize `nullptr`?

File: lib/Differentiator/VisitorBase.cpp [hunk 5]
```diff
@@ -379,9 +379,8 @@
...
     if (T->isVoidType())
       return nullptr;
-    if (T->isScalarType()) {
-      ExprResult Zero =
-          ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
+    else if (T->isScalarType() && !T->isPointerType()) {
+      ExprResult Zero = ConstantFolder::synthesizeLiteral(T, m_Context, 0);
       CastKind CK = m_Sema.PrepareScalarCast(Zero, T);
       return m_Sema.ImpCastExprToType(Zero.get(), T, CK).get();
     }
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/VisitorBase.cpp:386]: That's a good suggestion.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 3 above)
Comment (kchristin22) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1383]: True, it was a leftover from Vaibhav's PR. I will remove these lines.

File: lib/Differentiator/VisitorBase.cpp (hunk 4 above)
Comment (kchristin22) [reply lib/Differentiator/VisitorBase.cpp:382]: Right now, the pointers are initialized as an empty list, which is equivalent to `nullptr`, the same way clad tapes are initialized. I can alter them to be equal to `nullptr` if that's safer.

Comment (vgvassilev) [reply lib/Differentiator/VisitorBase.cpp:382]: Yes, that'd be more readable, too.

File: lib/Differentiator/VisitorBase.cpp [hunk 6]
```diff
@@ -379,11 +379,10 @@ namespace clad {
...
-    if (T->isScalarType()) {
+    if ((T->isScalarType() || T->isPointerType()) && !T->isReferenceType()) {
       ExprResult Zero =
-          ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
-      CastKind CK = m_Sema.PrepareScalarCast(Zero, T);
-      return m_Sema.ImpCastExprToType(Zero.get(), T, CK).get();
+          ConstantFolder::synthesizeLiteral(T, m_Context, /*val=*/0);
+      return Zero.get();
     }
     return m_Sema.ActOnInitList(noLoc, {}, noLoc).get();
```
Comment (vgvassilev) [inline lib/Differentiator/VisitorBase.cpp:387]: We should probably sink `ActOnInitList` into `synthesizeLiteral` but let's do it outside of this pr.

File: test/Gradient/FunctionCalls.C [hunk 7]
```diff
@@ -323,8 +323,8 @@ double fn8(double x, double y) {
 // CHECK-NEXT:     double _t1 = std::tanh(1.);
 // CHECK-NEXT:     double _t0 = std::max(1., 2.);
 // CHECK-NEXT:     {
-// CHECK-NEXT:         double _r0 = 0;
-// CHECK-NEXT:         char _r1 = 0;
+// CHECK-NEXT:         double _r0 = 0.;
+// CHECK-NEXT:         char _r1 = 0i8;
```
Comment (vgvassilev) [inline test/Gradient/FunctionCalls.C:327]: That will fail to compile I think.

File: test/Gradient/STLCustomDerivatives.C [hunk 8]
```diff
@@ -272,94 +272,94 @@ int main() {
 // CHECK-NEXT:     {
 // CHECK-NEXT:         res = _t25;
 // CHECK-NEXT:         double _r_d6 = _d_res;
-// CHECK-NEXT:         {{.*}} _r10 = 0;
+// CHECK-NEXT:         size_type _r10 = {{0U|0UL}};
 // CHECK-NEXT:         {{.*}}class_functions::operator_subscript_pullback(&_t26, 0, _r_d6, &_d_vec, &_r10);
-// CHECK-NEXT:         {{.*}} _r11 = 0;
+// CHECK-NEXT:         size_type _r11 = {{0U|0UL}};
```
Comment (vgvassilev) [inline test/Gradient/STLCustomDerivatives.C:277]: That will not compiler. We probably mean `std::vector<...>::size_type`. I do not think that's a problem for this PR.

**Review by vgvassilev — COMMENTED** [review]
We are almost there. See this one comment.

File: lib/Differentiator/VisitorBase.cpp (hunk 6 above)
Comment (kchristin22) [reply lib/Differentiator/VisitorBase.cpp:387]: I think it makes sense to be in `getZeroInit` as we initialize with an empty list, while synthesizeLiteral is used in other places as well

File: test/Gradient/STLCustomDerivatives.C (hunk 8 above)
Comment (kchristin22) [reply test/Gradient/STLCustomDerivatives.C:277]: I'll open an issue

File: test/Gradient/FunctionCalls.C (hunk 7 above)
Comment (kchristin22) [reply test/Gradient/FunctionCalls.C]: Yes indeed, fixed

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**